### PR TITLE
[WIP] Decouple managedTransport and improve transport handling logic.

### PR DESCRIPTION
**Pull request by evanlinjin**
Related to #443, #507 and #271

## Changes
* Fix transport re-establishment logic.
* Moved much of the logic from `transport.Manager` to `managedTransport` to improve clarity.
* Settlement Handshake: only the responder updates the status to transport discovery.
* Removed the ability to register 'private' transport.
* Better use of `context.Context` throughout the codebase.
* Added timeouts to `setup` library to avoid SetupNode hanging.
* Changed various golang examples to proper tests.
* Cleaned up some logging.
* Cleaned up some tests.
* TODO: Some `router` tests are failing and have been commented out. These need to be fixed at a later stage.

### CODE REVIEW

**Comment by evanlinjin** [comment]
I'm thinking about the following renaming:
- `Transport` -> `Connection` (ideally we can just use `net.Conn`).
- `ManagedTransport` -> `Transport`.

**Comment by evanlinjin** [comment]
Transport logic is still faulty.

File: pkg/router/route_manager.go [hunk 1]
```diff
@@ -35,6 +35,10 @@ func (rm *routeManager) GetRule(routeID routing.RouteID) (routing.Rule, error) {
 		return nil, errors.New("unknown RouteID")
 	}
 
+	if len(rule) < 13 {
```
Comment (nkryuchkov) [inline pkg/router/route_manager.go]: What do you think of extracting `13` to a constant?

File: pkg/setup/node.go [hunk 2]
```diff
@@ -67,14 +67,17 @@ func (sn *Node) Serve(ctx context.Context) error {
...
+			if err := sn.serveTransport(ctx, tp); err != nil {
 				sn.Logger.Warnf("Failed to serve Transport: %s", err)
 			}
 		}(tp)
 	}
 }
 
-func (sn *Node) serveTransport(tr transport.Transport) error {
+func (sn *Node) serveTransport(ctx context.Context, tr transport.Transport) error {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
```
Comment (nkryuchkov) [inline pkg/setup/node.go]: What do you think of extracting a timeout constant here?

File: pkg/setup/protocol.go [hunk 3]
```diff
@@ -130,35 +132,53 @@ func DeleteRule(p *Protocol, routeID routing.RouteID) error {
...
+func LoopClosed(ctx context.Context, p *Protocol, ld routing.LoopData) error {
 	if err := p.WritePacket(PacketLoopClosed, ld); err != nil {
 		return err
 	}
-	return readAndDecodePacket(p, nil)
+	return readAndDecodePacketWithTimeout(ctx, p, nil)
+}
+
+func readAndDecodePacketWithTimeout(ctx context.Context, p *Protocol, v interface{}) error {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
```
Comment (nkryuchkov) [inline pkg/setup/protocol.go]: I'd extract a constant here too.

File: pkg/transport/managed_transport.go [hunk 4]
```diff
@@ -2,107 +2,351 @@ package transport
...
+			break
+		}
+		select {
+		case <-mt.done:
+			return nil, ErrNotServing
+		case <-mt.connCh:
+		}
+	}
+
+	h := make(routing.Packet, 6)
```
Comment (nkryuchkov) [inline pkg/transport/managed_transport.go]: Can we extract `6` here?

**Review by nkryuchkov — APPROVED** [review]
Looks awesome! I'd extract some constant values to improve the readability, but it's up to you to decide.

**Pull request merged**